[3.13] gh-128679: Fix tracemalloc.stop() race conditions

vstinner opened the pull request:
tracemalloc_alloc(), tracemalloc_realloc(), PyTraceMalloc_Track(), PyTraceMalloc_Untrack() and _PyTraceMalloc_TraceRef() now check tracemalloc_config.tracing after calling TABLES_LOCK().

_PyTraceMalloc_Stop() now protects more code with TABLES_LOCK(), especially setting tracemalloc_config.tracing to 1.

Add a test using PyTraceMalloc_Track() to test tracemalloc.stop() race condition.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-128679 -->
* Issue: gh-128679
<!-- /gh-issue-number -->

vstinner committed:
gh-128679: Fix tracemalloc.stop() race conditions

tracemalloc_alloc(), tracemalloc_realloc(), PyTraceMalloc_Track(),
PyTraceMalloc_Untrack() and _PyTraceMalloc_TraceRef() now check
tracemalloc_config.tracing after calling TABLES_LOCK().

_PyTraceMalloc_Stop() now protects more code with TABLES_LOCK(),
especially setting tracemalloc_config.tracing to 1.

Add a test using PyTraceMalloc_Track() to test tracemalloc.stop()
race condition.

ZeroIntensity commented on Python/tracemalloc.c:
```diff
@@ -963,6 +973,10 @@ _PyTraceMalloc_Stop(void)
     if (!tracemalloc_config.tracing)
         return;
 
+    // Lock to synchronize with tracemalloc_free() which checks
```
We need to lock a little bit earlier for the read above.

vstinner replied:
Oh right, fixed.

ZeroIntensity commented on Python/tracemalloc.c:
```diff
@@ -1376,16 +1395,21 @@ _PyTraceMalloc_TraceRef(PyObject *op, PyRefTracerEvent event, void* Py_UNUSED(ig
     int res = -1;
 
     TABLES_LOCK();
```
We also read `tracing` without a lock before this.

vstinner replied:
Checking  `tracing` without the lock is a workaround for the Python finalization. TABLES_LOCK() can no longer be called after _PyTraceMalloc_Fini() (the lock is destroyed), whereas _PyTraceMalloc_TraceRef() is still called after _PyTraceMalloc_Fini().

`tracing` is checked again once we acquired the lock.

The issue was fixed properly in the main branch by calling `(void)PyRefTracer_SetTracer(NULL, NULL);` in `_PyTraceMalloc_Stop()`. Since this change is designed to be backported to Python 3.12, I didn't include this fix.

Once this change is merged in 3.13 and 3.12, we can remove the unprotected `tracing` read in 3.13 in `_PyTraceMalloc_TraceRef()` by adding `(void)PyRefTracer_SetTracer(NULL, NULL);`.

ZeroIntensity commented on Python/tracemalloc.c:
```diff
@@ -1418,7 +1442,9 @@ _PyTraceMalloc_ClearTraces(void)
     }
     set_reentrant(1);
-    tracemalloc_clear_traces();
+    TABLES_LOCK();
```
Right above here we have a lockless read.

vstinner replied:
Fixed

ZeroIntensity reviewed (commented):
There's a couple other (unmodified) functions where I'm seeing lockless reads:

- `tracemalloc_get_traceback`
- `_PyMem_DumpTraceback`
- `_PyTraceMalloc_GetTraces`
- `_PyTraceMalloc_GetTracedMemory`
- `_PyTraceMalloc_ResetPeak`

vstinner committed:
Fix the race condition in more functions

vstinner committed:
Lock also in _PyTraceMalloc_IsTracing()

vstinner committed:
_PyTraceMalloc_Start(): lock before reading tracing

vstinner committed:
Call _PyTraceMalloc_Init() at Python startup

So TABLES_LOCK() can be called in _PyMem_DumpTraceback() even if
tracemalloc is not tracing.

ZeroIntensity reviewed (approved):
LGTM. With the exception of a few special cases that you've mentioned, everything is properly locked! Due to the changes to the lifecycle and the size of the change, I think we should run buildbots before merging.

vstinner committed:
gh-128679: Skip test_tracemalloc_track_race() on debug build (#128988)

There is a race condition between PyMem_SetAllocator() and
PyMem_RawMalloc()/PyMem_RawFree(). While PyMem_SetAllocator() write
is protected by a lock, PyMem_RawMalloc()/PyMem_RawFree() reads are
not protected by a lock. PyMem_RawMalloc()/PyMem_RawFree() can be
called with an old context and the new function pointer.

On a release build, it's not an issue since the context is not used.
On a debug build, the debug hooks use the context and so can crash.

(cherry picked from commit 9bc19643ed54b2aa8558e87fd3a04e8bd1ea7fb2)
